### DISCUSSION TITLE
Support for iOS platform to use kvssink plugin

### DISCUSSION THREAD

**Discussion by sbelbin** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#1211] [General]
The GStreamer project supports Apple's iOS platform* such as iPhone and iPad devices whereas the AWS Kinesis Video streams producer SDK project (referred to as **this SDK** project) already supports Mac OS platform, but I haven't found any documentation/topic about building and deploying the AWS kvssink plugin for iOS platform.

>   _*The GStreamer provided package contains static libraries, without any dynamic libraries (*.dylib) with several _common_ GStreamer plugins built-in. The fact that GStreamer only provides static library becomes an issue that I'll explain later on._

For building this SDK project, I followed the guidelines along with option to build the GStreamer plugin (kvssink). This produces a dynamic library (libgstkvssink.dylib).

I use an iPad app to validate GStreamer in which I can provide a GStreamer pipeline using some of the _common_ GStreamer plugins to ensure that GStreamer on iPad works.

When I include the libgstkvssink.dyilb into the iPad and when I put the kvssink directive into the pipeline, the app fails because it lacks dependencies such libgstreamer-1.0.0.dylib & libgstbase-1.0.0.dylib.

> I'd anticipated this since the GStreamer provides an iOS package along with an XCode framework thats contains static libraries without dynamic libraries.

My questions for this discussion group:
- Has anyone faced this issue of using kvssink GStreamer plugin on iOS platform?
- If so how did your resolve it?

I'm reaching out to this group before exploring into other avenues. One of which is to incorporate the kvssink as a GStreamer's **subproject** is built into GStreamer as a static library. Thus GStreamer iOS package contains the kvssink as static libraries alongside the other _common_ plugins of the other subprojects.

Regards,
Steven Belbin

**Comment by sbelbin** [comment]
I've progressed on this matter so that it can be deployed and running on iOS devices on ARM64 (Apple M processors), see the recipe below.

At this point my test app on the iPad starts where it defines a pipeline in which the sink is to an AWS Kinesis video stream that is already created (which I've already verified on my MAC OSx). However, on the iPad it fails due to certificate issues see below:

`
kvssink stream-name=****************** storage-size=128 aws-region=****** access-key=**************** secret-key=********************************
`

`
2024-12-10 15:27:26 [0x16c93f000] WARN - curlCompleteSync(): [******************] curl perform failed for url https://kinesisvideo.*****.amazonaws.com/describeStream with result SSL peer certificate or SSH remote key was not OK: 
`

I'm looking into possibly using the `iot-certificate`options to the `kvssink` plugin & provide the necessary certificates file.

However, this seems to me as heavy handed solution. I'd rather find another option.

Any suggestions?

Regards,
Steven

**Recipe to get AWS Video Stream AWS Video Stream GStreamer plugin build and deployed for iOS platforms**

Build the AWS Video Stream GStreamer plugin as a static library **libgstkvssink.a** as well as all of its dependencies such as:
- libKinesisVideoProducer.a
- libcproducer.a
- libkvspic.a
- libkvspicUtils.a
- libkvsCommonCurl.a
- libcurl.a
- liblog4cplus.a 

Adapt the XCode project(s) as to reference as to provide the library folder path(s) to those static libraries and directives for each static library.

After which, provide the following code to declare & register the **kvssink** plugin. This is usually done in the **gst_ios_init.m** file, see GStreamer's iOS examples.

```C++
...

#if defined(GST_IOS_PLUGIN_KVSSINK)
GST_PLUGIN_STATIC_DECLARE(kvssink);
#endif


void
gst_ios_init (void)
{
...
#if defined(GST_IOS_PLUGIN_KVSSINK)
    GST_PLUGIN_STATIC_REGISTER(kvssink);
#endif
...
}
```

**Comment by sbelbin** [comment]
I've troubleshooting the issue CA certificates when running on iOS devices. Typically those certificates are stored in Keychain but since iOS 13.0 and the Secure-Transport capabilities is deprecated and to use Network framework instead and thus libCURL doesn't have access the CA certificates stored in the Keychain.

> I've reviewed the CURL/libCURL project's discussion on GitHub project and elsewhere and no one mentions a **new** build option to replace **--with-secure-transport** when building CURL/libCURL. Likewise, no one has stated about working on adapting the implementation to use Network framework instead of Secure-Transport.
>
>I'll report it to the CURL project hoping that  support to use Network framework gets added or another means to access the iOS Keychain.

As a workaround, I proceeded to bundle a CA certificates file with the app and for the app to set an environment variable which the AWS Kinesis video streamer code when preparing a CURL request when the **certPath** variable is undefined or empty.

 **.../libkvscproducer/kvscproducer-src/src/source/Request.c**
```c++
STATUS createCurlRequest(HTTP_REQUEST_VERB curlVerb, PCHAR url, PCHAR body, STREAM_HANDLE streamHandle, PCHAR region, UINT64 currentTime,
                         UINT64 connectionTimeout, UINT64 completionTimeout, UINT64 callAfter, PCHAR certPath, PAwsCredentials pAwsCredentials,
                         PCurlApiCallbacks pCurlApiCallbacks, PCurlRequest* ppCurlRequest)
{
...
    PCHAR caCertPath = (certPath != NULL && certPath[0] != 0) ? certPath : getenv(CACERT_PATH_ENV_VAR);

    if (caCertPath != NULL) {
        STRNCPY(pCurlRequest->requestInfo.certPath, caCertPath, MAX_PATH_LEN);
    }
...
```

> **BTW:** The C++ sampler code that are provided AWS Kinesis Video Producer SDK for C++ uses an environment variable (which is a macro to **'AWS_KVS_CACERT_PATH'**) which contains the file path of CA certificate file.
>
>>I'd prefer a solution of passing a **cert-path** property via the **kvssink** plugin.
>>However, _without_ all the requisites of declaring of the **iot-certificate** property.
>
>Anyways, this workaround of using an environment variable to allow me to continue._



Unfortunately, I've arrived at the next impediment (issue) where AWS Kinesis Video Producer SDK C++ fails to start the sink. This time the error message is cryptic (at best) with logging set to trace levels.

>[ERROR] [13-12-2024 18:28:43:567.683 UTC] createKinesisVideoStream(): operation returned status code: **0x15000020**

The value **0x15000020** without a message isn't significant about where to proceed in investigating the cause or causes.

I'd appreciate any insight for those whom support this project or from some that has become familiar with the implementation details on how to troubleshoot such issues with this project.

Regards,
Steven Belbin


**Logged messages...**
```
...
[TRACE] [13-12-2024 18:28:43:563.593 UTC] stepStateMachine(): [CLIENT] State Machine - Current state: 0x0000000000000010, Next state: 0x0000000000000040, Current local state retry count [0], Max local state retry count [0], State transition wait time [0] ms
[DEBUG] [13-12-2024 18:28:43:563.630 UTC] clientReadyHandler invoked
[TRACE] [13-12-2024 18:28:43:563.670 UTC] createKinesisVideoClientSync(): Awaiting for the Kinesis Video Client to become ready...
[TRACE] [13-12-2024 18:28:43:563.712 UTC] createKinesisVideoClientSync(): Kinesis Video Client is Ready.
[INFO ] [13-12-2024 18:28:43:563.856 UTC] Try creating stream for ********************
[INFO ] [13-12-2024 18:28:43:563.969 UTC] Creating Kinesis Video Stream ********************
[INFO ] [13-12-2024 18:28:43:564.021 UTC] createKinesisVideoStream(): Creating Kinesis Video Stream.
[INFO ] [13-12-2024 18:28:43:564.061 UTC] logStreamInfo(): SDK version: 3cf67319ae7baeeb940bf06eabdfdec127ebbcfe
[DEBUG] [13-12-2024 18:28:43:564.107 UTC] logStreamInfo(): Kinesis Video Stream Info
[DEBUG] [13-12-2024 18:28:43:564.146 UTC] logStreamInfo(): 	Stream name: ******************** 
[DEBUG] [13-12-2024 18:28:43:564.179 UTC] logStreamInfo(): 	Streaming type: STREAMING_TYPE_REALTIME 
[DEBUG] [13-12-2024 18:28:43:564.248 UTC] logStreamInfo(): 	Content type: video/h264 
[DEBUG] [13-12-2024 18:28:43:564.286 UTC] logStreamInfo(): 	Max latency (100ns): 600000000
[DEBUG] [13-12-2024 18:28:43:564.329 UTC] logStreamInfo(): 	Fragment duration (100ns): 20000000
[DEBUG] [13-12-2024 18:28:43:564.363 UTC] logStreamInfo(): 	Key frame fragmentation: Yes
[DEBUG] [13-12-2024 18:28:43:564.393 UTC] logStreamInfo(): 	Use frame timecodes: Yes
[DEBUG] [13-12-2024 18:28:43:564.424 UTC] logStreamInfo(): 	Absolute frame timecodes: Yes
[DEBUG] [13-12-2024 18:28:43:564.455 UTC] logStreamInfo(): 	Nal adaptation flags: 0
[DEBUG] [13-12-2024 18:28:43:564.490 UTC] logStreamInfo(): 	Average bandwidth (bps): 4194304
[DEBUG] [13-12-2024 18:28:43:564.525 UTC] logStreamInfo(): 	Framerate: 25
[DEBUG] [13-12-2024 18:28:43:564.555 UTC] logStreamInfo(): 	Buffer duration (100ns): 1200000000
[DEBUG] [13-12-2024 18:28:43:564.585 UTC] logStreamInfo(): 	Replay duration (100ns): 400000000
[DEBUG] [13-12-2024 18:28:43:564.622 UTC] logStreamInfo(): 	Connection Staleness duration (100ns): 600000000
[DEBUG] [13-12-2024 18:28:43:564.652 UTC] logStreamInfo(): 	Store Pressure Policy: 1
[DEBUG] [13-12-2024 18:28:43:564.682 UTC] logStreamInfo(): 	View Overflow Policy: 1
[DEBUG] [13-12-2024 18:28:43:564.712 UTC] logStreamInfo(): 	Allow stream creation: Yes
[DEBUG] [13-12-2024 18:28:43:564.744 UTC] logStreamInfo(): 	Segment UUID: NULL
[DEBUG] [13-12-2024 18:28:43:564.782 UTC] logStreamInfo(): 	Frame ordering mode: 0
[DEBUG] [13-12-2024 18:28:43:564.819 UTC] logStreamInfo(): Track list
[DEBUG] [13-12-2024 18:28:43:564.864 UTC] logStreamInfo(): 	Track id: 1
[DEBUG] [13-12-2024 18:28:43:564.903 UTC] logStreamInfo(): 	Track name: kinesis_video
[DEBUG] [13-12-2024 18:28:43:564.942 UTC] logStreamInfo(): 	Codec id: V_MPEG4/ISO/AVC
[DEBUG] [13-12-2024 18:28:43:564.976 UTC] logStreamInfo(): 	Track type: TRACK_INFO_TYPE_VIDEO
[DEBUG] [13-12-2024 18:28:43:565.007 UTC] logStreamInfo(): 	Track cpd: NULL
[TRACE] [13-12-2024 18:28:43:565.322 UTC] stepStateMachine(): [STREAM] State Machine - Current state: 0x0000000000000001, Next state: 0x0000000000000002, Current local state retry count [0], Max local state retry count [5], State transition wait time [0] ms
[TRACE] [13-12-2024 18:28:43:565.407 UTC] stepStateMachine(): [CLIENT] State Machine - Current state: 0x0000000000000040, Next state: 0x0000000000000040, Current local state retry count [1], Max local state retry count [0], State transition wait time [0] ms
[ERROR] [13-12-2024 18:28:43:567.683 UTC] createKinesisVideoStream(): operation returned status code: 0x15000020
[ERROR] [13-12-2024 18:28:43:567.737 UTC] createKinesisVideoStreamSync(): operation returned status code: 0x15000020
[ERROR] [13-12-2024 18:28:43:567.788 UTC]  Unable to create Kinesis Video stream. ******************** Error status: 0x15000020
[INFO ] [13-12-2024 18:28:43:569.321 UTC] Freeing Kinesis Video Stream for ********************
[INFO ] [13-12-2024 18:28:43:569.382 UTC] freeKinesisVideoStream(): Freeing Kinesis Video stream.
...
```

**Reply by sirknightj** [reply]
The status code `0x15000020` refers to `STATUS_INVALID_CA_CERT_PATH`.

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/5fb356e484d734ce3650c751a4469985c488bc1f/src/include/com/amazonaws/kinesis/video/common/Include.h#L49
https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/producer-sdk-errors.html

There are 2 KVS-owned dependent libraries of this C++ producer:
- C producer: https://github.com/awslabs/amazon-kinesis-video-streams-producer-c
- PIC: https://github.com/awslabs/amazon-kinesis-video-streams-pic

If you're not able to find the definition in this repo, then it's probably in the other two repos.
It's pretty much indicating here that the CA cert path needs to end with `.pem`. https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/5fb356e484d734ce3650c751a4469985c488bc1f/src/source/Response.c#L167-L171

Try printing out the string and make sure it's also properly null terminated.

Regarding IoT certs, you can refer to https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-iot.html, those are used to obtain temporary access key, secret key, eventually once you get past PoC stage, when you deploy it to actual devices in the field and you don't want to hard-code credentials in your application.

**Reply by sirknightj** [reply]
You may also want to double check what these are set to, if any.

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/5fb356e484d734ce3650c751a4469985c488bc1f/src/include/com/amazonaws/kinesis/video/common/Include.h#L254-L258

**Reply by sbelbin** [reply]
I found & resolved which was due to passing a file path of **'ca-bundle.crt'** rather than a file path with a **'.pem'** file extension as you've stated.

I appreciate your reply & those references links & these error codes will be handy (hopefully not too often).

With that quick correction in place the stream is ready.
```
[TRACE] [13-12-2024 22:07:26:603.696 UTC] createKinesisVideoStreamSync(): Kinesis Video Stream is Ready.
Stream is ready
```

As before the next impediment is a failure when transitioning the GStreamer pipeline to a PLAYING state due to an internal error.

Here is an example pipeline that I'm using.
```
videotestsrc is-live=true ! video/x-raw, framerate=10/1, width=640, height=480 ! vtenc_h264_hw allow-frame-reordering=FALSE realtime=TRUE max-keyframe-interval=45 bitrate=500 ! h264parse ! video/x-h264, stream-format=avc, alignment=au, profile=baseline ! kvssink stream-name=************ storage-size=128 aws-region=******** access-key=************** secret-key=***************** log-config=*****/logs/aws-kvs.log
```

I learnt that the **kvssink** plugin accepts **h264** & **h265** as well as **avc** thus I'm defining the pipeline accordingly. 
 
I use GStreamer's test video signal as the source as a baseline to verify that video is streamed to AWS Kinesis video stream. I've also tried with the GStreamer plugin for Apple devices **avfvideosrc** as the source.

> If you something wrong in the pipeline I'd appreciate the assist.

Regards,
Steven Belbin

**Reply by sirknightj** [reply]
Hi,

I tried out the pipeline you shared above and it's is working fine on my mac:
```sh
gst-launch-1.0 videotestsrc is-live=true ! video/x-raw, framerate=10/1, width=640, height=480 ! vtenc_h264_hw allow-frame-reordering=FALSE realtime=TRUE max-keyframe-interval=45 bitrate=500 ! h264parse ! video/x-h264, stream-format=avc, alignment=au, profile=baseline ! kvssink stream-name=demo-stream
```

Note that it does take a little bit to show up in the console playback though. It can load quicker if you reduce the max-key-frame interval in the encoder.

I also tried using this pipeline on my Mac, also with success uploading.

```sh
gst-launch-1.0 avfvideosrc  !  vtenc_h264_hw allow-frame-reordering=FALSE realtime=TRUE max-keyframe-interval=45 bitrate=500 ! h264parse ! kvssink stream-name="hi"
```

I'm not sure that the `_hw` variant of `vtenc_h264` is available on iPhones, at least, according to the documentation: https://gstreamer.freedesktop.org/documentation/applemedia/vtenc_h264_hw.html?gi-language=c
> Apple VideoToolbox H264 HW-only encoder (only available on macOS at present).

You can try using the non-hardware version instead: https://gstreamer.freedesktop.org/documentation/applemedia/vtenc_h264.html?gi-language=c 

Other than that, I don't see anything else obvious in your pipeline. Can you share the error you're seeing?

**Reply by sbelbin** [reply]
HI @sirknightj ,

As you've mention support & GStreamer capabilities running on a MacOS (Sequoia 15.0+) tends to be more complete than with its iOS (18.0+) counterpart.

I've got a GStreamer pipeline in which the iPad device to stream video to a video stream endpoint on the AWS Kinesis Video stream. That pipeline consists of using **vtenc_h264** + ***h264parse** in which their properties are set to their default settings.

```
avfvideosrc ! vtenc_h264 ! h264parse ! kvssink ...
```

When playing back the video using the **Media playback** of that video stream's AWS Kinesis Video page, it reports:

> **Fragment decoding error**
> **Details**
>------------------------
>Codec ID:
>avc1.4d003c (H.264 Main@6.0)
>Fragment number: 91343852333181492313931724237018968906270576110
>Producer timestamp: December 16, 2024 8:50:46 AM
>Size: 2.8 MB

To resolve this issue(s), I'll have to adjust the settings provided in the pipeline for the **vtenc_h264** + ***h264parse** and probably **kvssink** as reduce these occurrences during playback of the video stream from AWS Video Kinesis.

Once I find those settings (that is good enough for my use case) I'll post it here.

None the less, it's good seeing video being streamed from an iPad device to a video stream running on AWS Kinesis Video service.

I appreciate your support and guidance on this matter.

Regards,
Steven Belbin

P.S.

It's also good to confirm that the **Amazon Kinesis Video Streams producer SDK C++ project** can be built to run on iOS devices.

Albeit that is does require tweaks to make files & an implementation workaround so that **libCurl** has access the CA certificates in order to authenticate the AWS Kinesis Video service endpoint.
